### PR TITLE
Improve Node.view() method (closes #1238 and #1528)

### LINKED ISSUE #1238: make node.view() return a string
At the moment `node.view()` prints directly to `stdout`. It would be nicer if it returned a string which could then be output as required. This has come up as I've been using logging in our adjoint script and capturing the output from `view()` in order to log it is a pain.

Does anyone think this is a bad idea? It will obviously affect any scripts that use view().

### LINKED ISSUE #1528: Options to reduce indentation and remove colours in view() method
The doc strings for two transformations are longer than 80 chars and therefore pycodestyle complains. The reason the are longer is that they provide the expected output from the view() method in an example which both shows how this works to users but can also be tested by doctest.

A simple solution would be to tell the view() method to reduce the indentation from 4 spaces to 3. However, this is currently not an option. The doctest also fails because the output is marked up with colours. Having the option to remove that as well would be useful.

### PR DESCRIPTION
**Pull request by rupertford**
Adds support for changing the indentation depth, switching off coloured text and returns a string rather than printing to stdout.

### CODE REVIEW

**Comment by rupertford** [comment]
This looks ready for review to me. Probably one for @arporter or @sergisiso

**Review by arporter — CHANGES_REQUESTED** [review]
Looks good - this is a nice improvement.
Just minor things and places where you've not updated the notebooks:

    $ find . -name \*.ipynb -exec grep -H 'view[(]' {} \;
    ./nemo/nemo_example1.ipynb:    "The `view()` method provides a textual view of the PSyIR  tree:"
    ./nemo/nemo_example1.ipynb:    "schedule.view()"
    ./nemo/nemo_example2.ipynb:    "schedule.view()"
    ./nemo/nemo_example4.ipynb:    "schedule.view()"
    ./nemo/nemo_example4.ipynb:    "schedule.view()"
    ./nemo/nemo_example3.ipynb:    "schedule.view()"
    ./nemo/nemo_example3.ipynb:    "schedule.view()"
    ./nemo/nemo_example3.ipynb:    "schedule.view()"

    $ find . -name \*.rst -exec grep -H 'view[(]' {} \;
    ./doc/developer_guide/psyir.rst:   ``view()`` methods.
    ./doc/developer_guide/psyir.rst:    >>> schedule.view()
    ./doc/user_guide/profiling.rst:    schedule.view()
    ./doc/user_guide/profiling.rst:    schedule.view()
    ./doc/user_guide/psyke.rst:  schedule.view()
    ./doc/user_guide/psyke.rst:  schedule.view()
    ./doc/user_guide/getting_going.rst:By default, the ``view()`` method available on any PSyIR (PSyclone
    ./doc/user_guide/psy_layer.rst:in two ways. First the `view()` method outputs textual information about
    ./doc/user_guide/psy_layer.rst:   >>> schedule.view()
    ./doc/user_guide/transformations.rst:    > schedule.view()
    ./doc/user_guide/transformations.rst:    > schedule.view()

    $ find . -name \*.md -exec grep -H 'view[(]' {} \;
    ./tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/README.md:`schedule.view()` prints out a view of the whole schedule). You will
    ./tutorial/practicals/LFRic/distributed_memory/3_overlapping_comms/README.md:    schedule.view()
    ./tutorial/practicals/LFRic/single_node/2_openacc/README.md:    schedule.view()
    ./tutorial/practicals/nemo/1_nemo_psyir/README.md:4. Use the `view()` method of one of these `InlinedKern` nodes to
    ./tutorial/practicals/nemo/1_nemo_psyir/README.md:   kernels[0].view()

There's a TODO 1238 in transformations.rst and tl2ad.py also has two mentions of 1238.

**Comment by arporter** [comment]
Forgot to say, all tests and examples are OK with compilation.
Changes are pylint clean.

File: src/psyclone/psyir/nodes/node.py [hunk 1]
```diff
@@ -48,9 +48,6 @@
 from psyclone.errors import GenerationError, InternalError
 from psyclone.psyir.symbols import SymbolError
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/node.py:50]: Removed.

File: src/psyclone/psyir/nodes/node.py [hunk 2]
```diff
@@ -797,36 +794,60 @@ def depth(self):
...
+            result = f"{full_indent}{_index}: {description}\n"
+        children_result_list = []
+        for idx, node in enumerate(self._children):
+            children_result_list.append(
+                node.view(
+                    depth=depth + 1, _index=idx, colour=colour, indent=indent))
+        result = result + "".join(children_result_list)
+        return result
 
     def list(self, indent=0):
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/psyir/nodes/node.py:852]: Removing it does not affect any tests so I've assumed it is OK.

File: src/psyclone/tests/psyir/nodes/node_test.py [hunk 3]
```diff
@@ -43,18 +43,18 @@
 import os
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/nodes/node_test.py:43]: Done.

File: src/psyclone/tests/psyir/nodes/range_test.py [hunk 4]
```diff
@@ -211,7 +211,7 @@ def test_range_str():
     assert erange.node_str(colour=False) == str(erange)
 
 
-def test_range_view(capsys):
+def test_range_view():
```
[the remark this answers is not in the document]
Comment (rupertford) [reply src/psyclone/tests/psyir/nodes/range_test.py:214]: OK.

**Comment by rupertford** [comment]
I can't believe I missed so many required changes from ***.view() to print(***.view()). Thanks for finding them. There were a load more in the examples directories too. You will see that there are a lot more modified files now (another 80 or so). I think I've got them all. In terms of referencing the associated issue, I could not find a reference in tl2ad.py.

**Comment by rupertford** [comment]
Ready for next review from @arporter

**Review by arporter — APPROVED** [review]
All looks good. I can't find any old uses of view() now.
Tests and examples are all OK with compilation.
Docs all build fine.

File: src/psyclone/psyad/tl2ad.py [hunk 5]
```diff
@@ -95,7 +95,7 @@ def generate_adjoint_str(tl_fortran_str, active_variables, create_test=False):
 
     # Apply any required transformations to the TL PSyIR
     logger.debug("Preprocessing")
-    preprocess_trans(tl_psyir)
+    preprocess_trans(tl_psyir, active_variables)
```
[the remark this answers is not in the document]
Comment (arporter) [reply src/psyclone/psyad/tl2ad.py:98]: Now that I've updated to master this has been resolved.

**Pull request merged**